Source/NSGridView.m

```diff
@@ -0,0 +1,1233 @@
+
+  [_rows removeObjectAtIndex: index];
+  [_cells removeObjectsInArray: objs];
+
+  [self _refreshCells];
+}
+
+- (NSGridColumn *) addColumnWithViews: (NSArray*)views
+{
+  return [self insertColumnAtIndex: [self numberOfColumns]
+                         withViews: views];
+}
+
+- (NSGridColumn *) _insertColumnAtIndex: (NSInteger)index withCells: (NSArray *)cells
+{
+  NSGridColumn *gc = [[NSGridColumn alloc] init];
+
+  // Insert the row and release...
+  [_columns insertObject: gc atIndex: index];
+  [gc setGridView: self];
+  RELEASE(gc);
+
+  NSUInteger i = 0;
+  NSUInteger pos = index;
+  
+  for(i = 0; i < [self numberOfRows]; i++)
+    {
+      NSGridCell *c = nil;
+      NSGridRow *row = [_rows objectAtIndex: i];
+
+      if (i > ([cells count] - 1))
+        {
+          c = AUTORELEASE([[NSGridCell alloc] init]);
+        }
+      else
+        {
+          c = [cells objectAtIndex: i];
+        }
+      
+      [c _setOwningRow: row];
+      [c _setOwningColumn: gc];
+      [_cells insertObject: c
+                   atIndex: pos + i * [self numberOfColumns]];
+    }
+  
+  // Refresh...
+  [self _refreshCells];
+  return gc;
+}
+
+- (NSGridColumn *) insertColumnAtIndex: (NSInteger)index withViews: (NSArray *)views
+{
+  NSRect f = [self _prototypeFrame];
+  NSMutableArray *cells = [NSMutableArray arrayWithCapacity: [views count]];
+  FOR_IN(NSView*, v, views)
+    {
+      NSGridCell *c = [[NSGridCell alloc] init];
+
+      [v setFrame: f];
+      [c setContentView: v];
+      [cells addObject: c];
+      RELEASE(c);
+    }
+  END_FOR_IN(views);
+  return [self _insertColumnAtIndex: index withCells: cells];
+}
+
+- (void) moveColumnAtIndex: (NSInteger)fromIndex toIndex: (NSInteger)toIndex
+{
+  NSArray *cells = [self _cellsForColumnAtIndex: fromIndex];
+  [self removeColumnAtIndex: fromIndex];
+  [self _insertColumnAtIndex: toIndex withCells: cells];
+}
+
+- (void) removeColumnAtIndex: (NSInteger)index
+{
+  NSArray *objs = [self _cellsForColumnAtIndex: index];
+
+  [_columns removeObjectAtIndex: index];
+  [_cells removeObjectsInArray: objs];
+
+  [self _refreshCells];
+}
+
+- (NSGridCellPlacement) xPlacement
+{
+  return _xPlacement;
+}
+
+- (void) setXPlacement: (NSGridCellPlacement)x;
+{
+  _xPlacement = x;
+}
+
+- (NSGridCellPlacement) yPlacement;
+{
+  return _yPlacement;
+}
+
+- (void) setYPlacement: (NSGridCellPlacement)y;
+{
+  _yPlacement = y;
+}
+
+- (NSGridRowAlignment) rowAlignment;
+{
+  return _rowAlignment;
+}
+
+- (void) setRowAlignment: (NSGridRowAlignment)a;
+{
+  _rowAlignment = a;
+}
+
+- (CGFloat) rowSpacing
+{
+  return _rowSpacing;
+}
+
+- (void) setRowSpacing: (CGFloat)f
+{
+  _rowSpacing = f;
+}
+
+- (CGFloat) columnSpacing
+{
+  return _columnSpacing;
+}
+
+- (void) setColumnSpacing: (CGFloat)f
+{
+  _columnSpacing = f;
+}
+  
+- (void) mergeCellsInHorizontalRange: (NSRange)hRange verticalRange: (NSRange)vRange
+{
+  NSLog(@"Method %@ unimplemented.", NSStringFromSelector(_cmd));
```

Comment:
I think what needs to be done here is to set the _mergeHead of all the affected cells (sparing the first one). The interesting question is how this should show up in view drawing. Most likely views where the cell is merged should not be displayed and the first cell of such an area should cover the whole range.

Reply:
I won't be implementing merging right now.  I need to understand how it works as it is doing funny things (I attached an example in a comment).